File: 240107-read-files/2-16-test-file.c
/*
The pizza delivery service processes its orders on the first come, first served basis. After a client has paid for his/her order, he/she is added to the online waiting list. After the pizza is ready, the information is passed to a courier, saved to a file, and removed from the waiting list.
Display the number of customers who are currently waiting for an order. (Hint: to implement the task in C, use the structure that describes the concept of an order, with the properties: visitor's name, delivery address, order, total cost. An order is written from the waiting list to the file when it is deleted).
*/

#define _CRT_SECURE_NO_WARNINGS  // for compatibility with classic functions 
#include <stdio.h> 
#include <stdlib.h> 
 
struct order  //creating a structure for an order
{ 
    char name[30]; 
    char address[30]; 
    char pizza[30]; 
    float cost; 
    int quantity; 
    float sum; 
}; 
 
typedef struct order order; 
 
typedef struct queue //creating a structure for the order queue
{ 
    order date;      //order details
    order* head; 
    order* tail; 
    struct queue* next; 
} queue; 
 
queue* head = NULL, * tail = NULL; 
void instructions(); //expected functionality
void insert();       //adding an item to the queue
void removex();      //removing an item from the queue
void quantity();     //counting orders in the queue
void print();        //display of all orders
void fprint();       //saving a completed (deleted) order from the queue
 
int main() 
{ 
    queue* start = NULL; 
    int choice; 
    instructions(); 
    scanf("%d", &choice); 
    while (choice != 0){ 
        switch (choice){ 
            case 1: 
                insert(); break; 
            case 2: 
                fprint(); 
                removex(); 
                break; 
            case 3: 
                quantity(); break; 
            case 4: 
            print(); break; 
            default: 
                printf("Invalid entry. Please try again\n"); break; 
        } 
        instructions(); 
        printf("?\n"); 
        scanf("%d", &choice); 
    } 
    return 0; 
} 
 
void instructions(void) 
{ 
    printf("\n\n1) Add order to the queue\n" 
                "2) Deleting an order from the queue \n" 
                "3) Count orders\n" 
                "4) Output on display\n" 
                "0) Exit the program\n"); 
}  
 
void removex() 
{ 
    queue* p; 
    if (head != NULL) 
    { 
        p = head; 
        head = head->next; 
        free(p); 
        printf(" The order deleted\n"); 
    } 
    else printf("No orders\n"); 
} 
 
 
void insert() { 
    queue* p = head; 
    if(!(p = (queue*)malloc(sizeof(queue))))  
    {	 
        puts("Not enough memory"); 
        getchar();  
        return; 
    } 
    printf("Enter the name\n"); 
    getchar(); 
    fgets(p->date.name,40,stdin); 
    printf("Enter the address:\n"); 
    rewind(stdin); 
    fgets(p->date.address,20,stdin); 
    printf("\n"); 
    printf("Enter the order\n"); 
    int finish=1; 
    p->date.sum=0; 
    while(finish!=0) 
    { 
        printf("Enter the name of the pizza: \n"); 
        rewind(stdin); 
        getchar(); 
        fgets( p->date.pizza,15,stdin); 
        printf("Enter  cost of the pizza:\n"); 
        scanf("%f", &p->date.cost); 
        printf("Enter the quantity\n"); 
        scanf("%d",&p->date.quantity); 
        p->date.sum += p->date.quantity*p->date.cost; 
        printf("Finish order 0-yes 1-no\n"); 
        scanf("%d",&finish); 
    } 
 
    printf(" Total amount  $%.2f\n",p->date.sum); 
    p->next = NULL; 
    if (head == NULL) 
    { 
        head = p; 
    } 
    else tail->next = p; 
    tail = p; 
} 
 
void quantity() 
{ 
    queue* p = head; 
    int count = 0; 
    if (head == NULL) 
        printf("No orders"); 
    else 
    { 
        while (p != NULL) 
        { 
            count++; 
            p = p->next; 
        } 
        printf("Number of clients %d\n", count); 
    } 
} 
 
void print() 
{ 
    int i=1; 
    queue* p = head; 
    if (head != NULL) 
    { 
        printf("______________________________________________\n"); 
        printf("|  Order \n"); 
        printf("______________________________________________\n"); 
        while (p != NULL) 
        { 
            printf("%d| Name %15s| Address %15s| Total amount %.2f", i, p->date.name, p->date.address,p->date.sum); 
            printf("______________________________________________\n"); 
            p = p->next; 
            i++; 
        } 
    } 
    else printf("No orders\n"); 
}  
 
 
void fprint() 
{  
    FILE* f; 
    f = fopen("order.txt", "a"); 
    queue* p = head;      
 
    fprintf(f, "______________________________________________\n"); 
    fprintf(f, " Name %15s| Address %15s| Total amount %.2f\n", p->date.name, p->date.address,p->date.sum); 
    fprintf(f, "______________________________________________\n"); 
    p = p->next; 
    fclose(f); 
} 

File: 240107-read-files/2-16-test-file.py
"""
The pizza delivery service processes its orders on the first come, first served basis. After a client has paid for his/her order, he/she is added to the online waiting list. After the pizza is ready, the information is passed to a courier, saved to a file, and removed from the waiting list. Display the number of customers who are currently waiting for an order. (Hint: to implement the task in C, use the structure that describes the concept of an order, with the properties: visitor's name, delivery address, order, total cost. An order is written from the waiting list to the file when it is deleted).

Expected functionality:

    Enter the info (customer's name, address, order, total cost).
    Write to file – delete the order data.
    Display the number of customers waiting for the order (customer's name, address, total cost).

"""


import collections 
 
def user_menu(): 
    return int(input("1) Add order to the queue\n" 
                     "2) Delete the order from the queue\n" 
                     "3) Count orders\n" 
                     "4) Output on display\n" 
                     "0) Exit the program\n" 
                     "?\n")) 
 
 
def add_order(queue_): 
    order_data = dict() 
    order_data["name"] = input("Enter the name: ") 
    order_data["address"] = input("Enter the address: ") 
    order_data["order"] = [] 
    total_amount = 0 
    while True: 
        pizza_data = dict() 
        pizza_data['name'] = input("Enter the name of the pizza: ") 
        pizza_data['cost'] = float(input("Enter the cost of the pizza: ")) 
        pizza_data['quantity'] = int(input("Enter the quantity: ")) 
        total_amount += pizza_data["cost"] * pizza_data["quantity"] 
        order_data["order"].append(pizza_data) 
        if not int(input("Finish order 0-yes 1-no\n")): 
            break 
    order_data["amount"] = total_amount 
    print("Total amount ${}".format(total_amount)) 
    queue_.append(order_data)  
 
 
def delete_order(queue_): 
    order = queue_.popleft()
    with open('my_order.txt', 'a') as fh:
        fh.write('_'*80 + '\n')
        fh.write("Name {}| Address {}| Total amount {}\n".format(
            order["name"],
            order["address"],
            order["amount"],
        ))
        fh.write('_'*80 + '\n')
 
 
def count_orders(queue_): 
    print("Number of clients {}".format(len(queue_))) 
 
 
def output_on_display(queue_): 
    temp_queue = queue_.copy() 
    counter = 0 
    print(len(temp_queue)) 
    while len(temp_queue): 
        order = temp_queue.popleft() 
        counter += 1 
        print('-'*80) 
        print("Order {}".format(counter)) 
        print("Name {}".format(order["name"])) 
        print("Address {}".format(order["address"])) 
        print("Total amount {}".format(order["amount"])) 
        print('-'*80) 
 
 
MENU_ITEMS = { 
    1: add_order, 
    2: delete_order, 
    3: count_orders, 
    4: output_on_display, 
} 
 
orders_queue = collections.deque() 
while True: 
    choice = user_menu() 
    if not choice: 
        break 
    MENU_ITEMS[choice](orders_queue)

